proof/invariant-abstract/ARM/ArchDetSchedAux_AI.thy

```diff
@@ -274,14 +274,22 @@ lemma perform_asid_control_invocation_pred_map_sc_refill_cfgs_of:
   by (wpsimp wp: delete_objects_pred_map_sc_refill_cfgs_of
            comb: hoare_drop_imp)
 
-lemma perform_asid_control_invocation_inactive_implies_zero_budget:
+lemma perform_asid_control_invocation_implies_zero_budget:
   "perform_asid_control_invocation aci
-   \<lbrace>\<lambda>s. pred_map inactive_scrc (sc_refill_cfgs_of s) p
+   \<lbrace>\<lambda>s. pred_map Q (sc_refill_cfgs_of s) p
         \<longrightarrow> pred_map (\<lambda>cfg. scrc_budget cfg = 0) (sc_refill_cfgs_of s) p\<rbrace>"
```

Comment:
Not suggesting to change these, just out of curiosity: would it make sense to pull the implication into the pred_map? We don't seem to be doing this in other proofs either, but it looks shorter to me (and equivalent).

Reply:
Yeah I think we could do that, and without too much of a change. Maybe it would make the lifting rules a little more annoying to deal with.

Reply:
I'm definitely not suggesting changing anything at this point, but does us doing this for quite a few of our conditions indicate that we actually want finer-grained projections?

Reply:
Oh when I saw Gerwin's comment, I thought he was referring to the statement of the new conjunct in `active_scs_valid`.

I didn't pay an awful lot of attention to these lemmas for `retype_region` etc. I think they're basically saying "if an active sched contexts exists after some deletion process occurred, then we didn't actually touch it" or "if we did retype it, it's now a default object". I am having a better look now to see if they can be improved or generalised.

@corlewis what sort of finger-grained projections were you thinking of? I think the `sc_refill_cfgs_of` projection works pretty well. It was used an awful lot for `valid_refills`

Reply:
I did have a quick look to see whether it might be straightforward to get rid of the doubling up of `pred_map` in these sorts of implications. It's quite annoying, actually. I think these rules would work fairly well with `hoare_vcg_imp_lift'` or `hoare_drop_imp`, especially with the schematic predicates that are in some of these rules.

As for whether we should get rid of this sort of thing in the statement of the invariants, I would say that that would make sense, too. But I'm growing to rather dislike `pred_map` and its many often contorted variants. My vote would be to rewrite a lot of these invariants with `opt_map` in some way.

Should I leave all this as is for now?